3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2017-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics
+package snowplow.enrich
+package common.enrichments.registry
+package pii
+
+// Scala
+import scala.collection.JavaConverters._
+import scala.collection.mutable.MutableList
+
+// Scala libraries
+import org.json4s
+import org.json4s.{DefaultFormats, JValue}
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.jackson.Serialization.write
+
+// Java
+import org.apache.commons.codec.digest.DigestUtils
+
+// Java libraries
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ArrayNode, ObjectNode, TextNode}
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath}
+import com.jayway.jsonpath.MapFunction
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+// Iglu
+import iglu.client.validation.ProcessingMessageMethods._
+import iglu.client.{SchemaCriterion, SchemaKey}
+
+// This project
+import common.ValidatedNelMessage
+import common.utils.ScalazJson4sUtils
+import common.outputs.EnrichedEvent
+
+/**
+ * Companion object. Lets us create a PiiPseudonymizerEnrichment
+ * from a JValue.
+ */
+object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+
+  implicit val json4sFormats = DefaultFormats
+
+  override val supportedSchema =
+    SchemaCriterion("com.snowplowanalytics.snowplow.enrichments", "pii_enrichment_config", "jsonschema", 2, 0, 0)
+
+  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[PiiPseudonymizerEnrichment] = {
+    for {
+      conf <- matchesSchema(config, schemaKey)
```

Comment:
Which alignment are you referring to in the above code?

Reply:
First two lines of `for`-comprehension. They are not aligned, while everything after `=` aligned.

Reply:
the `<-` token is aligned in an unbroken run but because there is an intervening `=` there are effectively three aligned blocks. Formatters are not as clever as to fully understand the language syntax I don't think. For all those cases one can use:
```
// format off
...
// format on
```

Reply:
Yep. That is exactly an argument to not force automated formatters - we're swapping one visual garbage with another.

Reply:
In my opinion you probably get better overall results using a formatter to ensure consistent readability when a project gets above a few KLOC than not, even if you occasionally have to manually manage the format. It does most of the work for you, ensures consistency, and minimises formatting work so that people can focus on structure and potential defects. Consistent formatting is not a self-cause the way I see it but a means to an end, and that end is to improve maintainability and better discern potential defects. If teams can minimise the time spent on tasks that can be automated, is that not a win?

Reply:
> In my opinion you probably get better overall results using a formatter

I never said anything against using a formatter. I said only about **forcing** formatting.

Reply:
I agree with @knservis - it's better to apply global formatting and then use `format off / on` flags around pieces which are beyond the formatter's capability. This is a non-argument in other language communities (see e.g. Golang).

I don't agree that `// format off and // format on` is visual garbage - if anything, it's helpful - it says "here be dragons": the syntactical complexity of this piece of code is beyond a computer's ability to visualize it well for a human. That's interesting information.

On automated formatting - think of it as cruise control on a car - just because you have to sometimes switch it off for a complicated piece of driving, nobody seriously thinks it's better not to have it at all.